pkg/kubelet/cm/device_plugin_handler.go

```diff
@@ -0,0 +1,257 @@
+		needed := int(v.Value())
+		glog.V(3).Infof("needs %d %s", needed, key)
+		if !deviceplugin.IsDeviceName(k) || needed == 0 {
+			continue
+		}
+		// Gets list of devices that have already been allocated.
+		// This can happen if a container restarts for example.
+		if h.allocatedDevices[key] == nil {
+			h.allocatedDevices[key] = newPodDevices()
+		}
+		devices := h.allocatedDevices[key].getDevices(string(pod.UID), container.Name)
+		if devices != nil {
+			glog.V(3).Infof("Found pre-allocated devices for container %q in Pod %q: %v", container.Name, pod.UID, devices.List())
+			needed = needed - devices.Len()
+		}
+		// Get Devices in use.
+		devicesInUse := h.allocatedDevices[key].devices()
+		glog.V(3).Infof("all devices: %v", h.allDevices[key].List())
+		glog.V(3).Infof("devices in use: %v", devicesInUse.List())
+		// Get a list of available devices.
+		available := h.allDevices[key].Difference(devicesInUse)
+		glog.V(3).Infof("devices available: %v", available.List())
+		if int(available.Len()) < needed {
+			return nil, fmt.Errorf("requested number of devices unavailable. Requested: %d, Available: %d", needed, available.Len())
+		}
+		allocated := available.UnsortedList()[:needed]
+		resp, err := h.devicePluginManager.Allocate(key, append(devices.UnsortedList(), allocated...))
+		if err != nil {
+			return nil, err
+		}
+		ret = append(ret, resp)
+		for _, device := range allocated {
+			// Update internal allocated device cache.
+			h.allocatedDevices[key].insert(string(pod.UID), container.Name, device)
+		}
+	}
+	return ret, nil
+}
+
+// devicesInUse returns a list of custom devices in use along with the
+// respective pods that are using them.
+func devicesInUse() map[string]*podDevices {
+	// TODO: gets the initial state from checkpointing.
+	return make(map[string]*podDevices)
+}
+
+// updateAllDevices gets all healthy devices.
+// TODO: consider to remove this function.
+func (h *DevicePluginHandlerImpl) updateAllDevices() {
+	h.allDevices = make(map[string]sets.String)
+	for key, devs := range h.devicePluginManager.Devices() {
+		h.allDevices[key] = sets.NewString()
+		for _, dev := range devs {
+			glog.V(3).Infof("insert device %s for resource %s", dev.ID, key)
+			h.allDevices[key].Insert(dev.ID)
+		}
+	}
+}
+
+// updateAllocatedDevices updates the list of GPUs in use.
+// It gets a list of active pods and then frees any GPUs that are bound to
+// terminated pods. Returns error on failure.
+func (h *DevicePluginHandlerImpl) updateAllocatedDevices(activePods []*v1.Pod) {
+	activePodUids := sets.NewString()
+	for _, pod := range activePods {
+		activePodUids.Insert(string(pod.UID))
+	}
+	for _, v := range h.allocatedDevices {
+		allocatedPodUids := v.pods()
+		podsToBeRemoved := allocatedPodUids.Difference(activePodUids)
+		glog.V(5).Infof("pods to be removed: %v", podsToBeRemoved.List())
+		v.delete(podsToBeRemoved.List())
+	}
+}
+
+// A simple stub implementation for DevicePluginHandler.
+type DevicePluginHandlerStub struct{}
```

Comment:
nit: Move the stub implementation to a separate file.

Reply:
done.